cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java

```diff
@@ -98,18 +113,33 @@ public RuntimeMonitor(ProgramRunId programRunId, CConfiguration cConf,
     this.gracefulShutdownMillis = cConf.getLong(Constants.RuntimeMonitor.GRACEFUL_SHUTDOWN_MS);
     this.topicsToRequest = new HashMap<>();
     this.requestKeyToLocalTopic = new HashMap<>();
+    this.dsFramework = datasetFramework;
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(
+        new SystemDatasetInstantiator(datasetFramework), new TransactionSystemClientAdapter(txClient),
+        NamespaceId.SYSTEM, Collections.emptyMap(), null, null)),
+      org.apache.tephra.RetryStrategies.retryOnConflict(20, 100)
+    );
     this.programFinishTime = -1L;
     this.lastProgramStateMessages = new ArrayList<>();
   }
 
   @Override
   protected void startUp() throws Exception {
     LOG.debug("Starting runtime monitor for program run {}.", programRunId);
-    // TODO initialize from offset table for a given programId
+
+    Set<String> topicConfigs = getTopicConfigs();
     for (String topicConfig : getTopicConfigs()) {
-      topicsToRequest.put(topicConfig, new MonitorConsumeRequest(null, limit));
       requestKeyToLocalTopic.put(topicConfig, getTopic(topicConfig));
     }
+
+    callWithRetries((Retries.Callable<Void, TransactionFailureException>) () -> {
```

Comment:
One problem with looping infinitely in the startUp method of a guava Service is that, it can potentially block the `stopAndWait` call on this service forever. This will be causing issue in shutting down cdap-master when there is problem with the underlying storage.